### ISSUE TITLE
Need a simple method somewhere for loading data programmatically for those of us who are writing code.

### ISSUE THREAD

**Issue by jdramsey** [cmu-phil/tetrad#1145]
The data loading method has become quite complicated. Need easy methods to load continuous, discrete, and mixed data that programmers can easily use. Here's an example I just did:

    private DataSet getDiscreteData(String path, Delimiter delimiter, char quoteChar) throws IOException {
        TabularColumnFileReader colReader = new TabularColumnFileReader(
                new File(path).toPath(),
                delimiter
        );

        colReader.setQuoteCharacter(quoteChar);

        DataColumn[] cols = colReader.generateColumns(new int[0], true);

        TabularDataReader reader = new TabularDataFileReader(
                new File(path).toPath(),
                Delimiter.COMMA
        );

        Data data = reader.read(cols, true);

        reader.setQuoteCharacter('\"');

        return (DataSet) DataConvertUtils.toDataModel(data);
    }

**Comment by kvb2univpitt** [comment]
@jdramsey Unfortunately, the data reader gets more complex as we add more functionality.  What made the data reader more complex than before was adding functions for interventional variables.  If you're not planning to manipulate columns or applying metadata, you can do the following:

```java
ContinuousTabularDatasetReader dataReader = new ContinuousTabularDatasetFileReader(Paths.get("data.csv"), Delimiter.COMMA);

DataSet dataset = (DataSet) DataConvertUtils.toDataModel(dataReader.readInData());
```
This is as simple as it gets.

**Comment by yuanzhou** [comment]
@jdramsey here is the data-reader documentation: https://github.com/bd2kccd/data-reader/blob/development/README.md

**Comment by jdramsey** [comment]
@kvb2univpitt What's wrong with just having a simple function like the one I suggested? Then, if the data reader method changes, you can just update this function...

This comes up a lot. People are just completely baffled as to how to use the data reader. It took me an hour this morning to sort through all of the code to figure it out.

**Comment by jdramsey** [comment]
I'm not saying there's anything amiss with the data reader. Surely not. After going through the code, I completely understand why. But for the person who's trying to use the data reader in code, the changes are hard to follow; it's more comfortable just to go back to the old data reader, which is buggy. I thought maybe just a few simple static methods in the DataUtils class would solve the problem completely.

Just like this:

DataSet dataSet = DataUtils.loadData(path, delimiter, quotechar);

I'm happy to add them if you don't want to, but only if you don't object.

**Comment by mglymour** [comment]
Either way, if the data reader code is going to continue to be housed in a separate repository, we definitely need an easily-findable link to that documentation on the Tetrad github, since data loading is pretty much one of the most fundamental functions of Tetrad, and if you can't figure out how to make it work, you can't do much of anything else.

**Comment by kvb2univpitt** [comment]
@mglymour @jdramsey Clark, we are actually planning on moving the data into Tetrad.  We think it's unnecessary to keep them separate.

> What's wrong with just having a simple function like the one I suggested? Then, if the data reader method changes, you can just update this function...

Joe, there's nothing wrong with what you're suggesting.  Please feel free to create a wrapper for what you need.  The data reader is flexible enough to code it up any way you like.  The example above was just to show a simple way to read in data with just two lines of code. 

> This comes up a lot. People are just completely baffled as to how to use the data reader. It took me an hour this morning to sort through all of the code to figure it out.

Sorry to hear that had to go through the code to figure it out.  I created a lot of examples on how to use the data reader in the unit tests.  Plus, we have documentations that Zhou created.  

>  But for the person who's trying to use the data reader in code, the changes are hard to follow; it's more comfortable just to go back to the old data reader, which is buggy.

I try my best to keep things as simple as I can.  Again, from the example I posted above, it's only two lines of code to read in a data file with default settings.  I don't know how it make it more simpler.

**Comment by cg09** [comment]
The important thing is that the software somewhere show the simple
directions very prominently.

**Comment by jdramsey** [comment]
@kvb2univpitt  Part of it is my problem. I had been using the continuous tubular reader, but then I needed to load a mixed variable dataset with it, and whatever I was doing using the mixed reader kept throwing exceptions. I dug down into it and found that it had actually read in all of the data, but when converting the data to a DataSet the continuousData and discreteData fields weren't populated. I'm not sure what I had done wrong, or else I'd make a simple report for you.I'll keep thinking about it.

So I ended up digging down into all of the GUI code to find where the data was actually being used and ended up with the version I sent earlier, which worked. I ended up loading it as discrete data.

I'm sure I just screwed up somewhere. I'm not all that bright.

**Comment by mglymour** [comment]
I would be very much in favor of wrapper methods that remain static even if the underlying code is changing.

**Comment by kvb2univpitt** [comment]
@jdramsey Hmm...I am not sure why that didn't work for me.  Can you try the following and if it works?

```java
MixedTabularDatasetReader dataReader = new MixedTabularDatasetFileReader(Paths.get("data.txt"), Delimiter.TAB, 4);
DataSet dataset = (DataSet) DataConvertUtils.toDataModel(dataReader.readInData());
```
You might have to change the delimiter, the file name, and the "number of discrete category."
